### PR TITLE
Add list, mintAndCreate assistant API

### PR DESCRIPTION
**Pull request by jiyoung-an**
### Summary
- Add list assistant API.
- Add mint and create assistant API.
- Lot of refactoring...
   - Rename folder `chat` -> `ai` / `common` -> `utils`.
   - Split utils to separate files. (util.ts, env.ts, path.ts, validator.ts).
   - Delete `BlockchainBase` class and update to inherit `FactoryBase` class.
   - Add `AinizeService` singleton utility class.
- Skip unit tests.

### CODE REVIEW

File: src/utils/util.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  return new Promise((resolve) => {
+    setTimeout(resolve, ms);
+  });
+}
+
+export function serializeEndpoint(endpoint: string) {
+  return endpoint.endsWith('/') ? endpoint.slice(0, -1) : endpoint;
+}
+
+export const exists = async (path: string, ain: Ain): Promise<boolean> => {
```
Comment (platfowner) [inline src/utils/util.ts:102]: minor: how about valueExists() as it's only about db values?

File: src/ai/assistant.ts [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+    const tokens = await this.getTokensByAddress(objectId, address);
+    const assistants = this.getAssistantsFromTokens(tokens);
+    const sorted = this.sortAssistants(assistants, order);
+
+    const total = sorted.length;
+    const items = sorted.slice(offset, offset + limit);
+
+    return { total, items };
+    // NOTE(jiyoung): example data
+    /**
```
Comment (platfowner) [inline src/ai/assistant.ts:227]: minor) how about just using '/*' as it's not a tsdoc comment?

File: src/ai/assistant.ts [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+
+    const { data } = await this.ainize.requestWithAuth<AssistantMinted>(this.ain, {
+      serverName,
+      opType,
+      data: body,
+    });
+
+    return data;
+    // NOTE(jiyoung): example data
+    /**
```
Comment (platfowner) [inline src/ai/assistant.ts:289]: ditto?

**Review by platfowner — APPROVED** [review]
LGTM with some minor comments.
Thanks!

File: src/utils/util.ts (hunk 1 above)
Comment (jiyoung-an) [reply src/utils/util.ts]: done! thanks

File: src/ai/assistant.ts (hunk 2 above)
Comment (jiyoung-an) [reply src/ai/assistant.ts]: thank you! i made mistake with this syntax

File: src/ai/assistant.ts (hunk 3 above)
Comment (jiyoung-an) [reply src/ai/assistant.ts]: done! thanks

File: src/ai/assistant.ts [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+      ...(metadata && !_.isEmpty(metadata) && { metadata }),
+    };
+
+    const { data } = await this.ainize.requestWithAuth<Assistant>(this.ain, {
+      serverName,
+      opType,
+      data: body,
+    });
+
+    if (!isAdmin) {
```
Comment (jiyoung-an) [inline src/ai/assistant.ts:87]: @platfowner if user is not admin, only response data is returned. (in this case, backend server do admin operation)
but if user is admin, they send transaction with response data.

File: src/ai/assistant.ts [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+      name,
+      instructions,
+      ...(description && { description }),
+      ...(metadata && !_.isEmpty(metadata) && { metadata }),
+    };
+    const info = {
+      id,
+      type: 'chat',
+      config,
+      createdAt: created_at,
```
Comment (jiyoung-an) [inline src/ai/assistant.ts:332]: @platfowner I added two fields to assistant schema in blockchain. (type, createdAt)
- type: 
   - 'chat', (in future, 'image', 'multi-modal', 'voice'..)
   - type of assistant is determined by the specialized functions of the model.
   - ref) https://huggingface.co/models
- createdAt:
   -  need for retrieving and sorting assistant list information from blockchain data.

File: src/ai/assistant.ts [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+      history: true,
+    };
+
+    const rules = {
+      write: `auth.addr === $user_addr || util.isAppAdmin('${appId}', auth.addr, getValue) === true`,
+      state: {
+        thread: {
+          gc_max_siblings: THREAD_GC_MAX_SIBLINGS,
+          gc_num_siblings_deleted: THREAD_GC_NUM_SIBLINGS_DELETED,
```
Comment (jiyoung-an) [inline src/ai/assistant.ts:341]: I added gc rule for thread too. (gc_max_siblings: 50, gc_num_siblings_deleted: 30)

File: src/constants.ts [hunk 7]
```diff
@@ -3,29 +3,27 @@ export const NODE_ENV = process.env.NODE_ENV;
...
 
 export const SUPPORTED_AINFT_STANDARDS = {
   721: '721',
-}
+};
 
-export const SERVICE_NAME_MAP = new Map<string, string>([
-  ['openai', 'ainize_openai']
-]);
+export const THREAD_GC_MAX_SIBLINGS = 50;
```
Comment (platfowner) [inline src/constants.ts:25]: minor) if they do not look too small to you, how about using smaller values, e.g. (20, 10) pair? @jiyoung-an

Comment (jiyoung-an) [reply src/constants.ts:25]: @platfowner i think 50 is too big too, so how about (30, 10)?

Comment (platfowner) [reply src/constants.ts:25]: 30 also looks big to me. :) If possible, keeping the value as small as possible would help the budget management. And it looks like more than 20 threads cannot be rendered at once on the left-side panel?

Comment (jiyoung-an) [reply src/constants.ts:25]: it's good!  I will be adjusted to (20, 10) in next pr! 
it seems appropriate to fetch 20 threads per call (similar rendered values)

Comment (platfowner) [reply src/constants.ts:25]: sounds great, thanks!

**Pull request merged**